web3walletcore/src/commonMain/kotlin/com/sonsofcrypto/web3walletcore/services/actions/ActionsService.kt

```diff
@@ -2,86 +2,64 @@ package com.sonsofcrypto.web3walletcore.services.actions
+    /** List of actions yet to be completed */
+    fun actions(): List<Action>
+    /** Mark action as completed */
+    fun markComplete(action: Action)
     /** Add a listener to the service */
-    fun addListener(listener: ActionsServiceListener)
+    fun addListener(listener: ActionsListener)
     /** Remove a listener from the service */
-    fun removeListener(listener: ActionsServiceListener)
-    /** Complete an action */
-    fun completeActionType(type: Action.Type)
-    /** List of actions yet to be completed */
-    fun outstandingActions(): List<Action>
+    fun removeListener(listener: ActionsListener)
 }
 
 class DefaultActionsService(
     private val store: KeyValueStore,
     private val networksService: NetworksService,
 ): ActionsService {
-    private var listeners = mutableListOf<ActionsServiceListener>()
-
-    override fun addListener(listener: ActionsServiceListener) {
-        listeners.add(listener)
-    }
-
-    override fun removeListener(listener: ActionsServiceListener) {
-        listeners.remove(listener)
-    }
 
-    override fun completeActionType(type: Action.Type) {
-        if (isActionCompleted(type)) return
-        store[type.identifier] = true
-        listeners.forEach { it.actionsUpdated() }
-    }
+    private var listeners: MutableList<WeakRef<ActionsListener>> = mutableListOf()
 
-    override fun outstandingActions(): List<Action> {
+    override fun actions(): List<Action> {
         val actions = mutableListOf<Action>()
-        confirmMnemonicAction?.let { if (!isActionCompleted(it.type)) { actions.add(it) } }
-        actions.add(themesAction)
-        actions.add(improvementProposalsAction)
+        if (!isCompleted(Action.ConfirmMnemonic)) {
+            actions.add(Action.ConfirmMnemonic)
+        }
+        actions.add(Action.Themes)
+        actions.add(Action.ImprovementProposals)
         return actions
     }
 
-    private fun isActionCompleted(type: Action.Type): Boolean = store[type.identifier] ?: false
-
-    private val confirmMnemonicAction: Action? get() {
-        val address = selectedWalletAddress() ?: return null
-        return Action(
-            ConfirmMnemonic(address),
-            "s",
-            "Mnemonic",
-            "Confirm your mnemonic",
-            "modal.mnemonic.confirmation",
-            false,
-            1
-        )
+    override fun markComplete(action: Action) {
+        store[actionKey(action)] = true
+        listeners.forEach { it.value?.actionsUpdated() }
     }
 
-    private fun selectedWalletAddress(): String? = networksService.wallet()?.id()
+    private fun isCompleted(action: Action): Boolean =
+        store.get<Boolean>(actionKey(action)) == true
+
+    private fun actionKey(action: Action): String = when (action) {
+        is Action.ConfirmMnemonic -> {
+            (action::class.simpleName ?: "") +
+                (networksService.wallet()?.id() ?: "")
+        }
+        else -> action::class.simpleName ?: "$action"
+    }
```

Comment:
No need for separete `Action.Type` and identifier. `action::class.simpleName` is same as `"\(SomeClass.self)"`
Incase of confirm mnemonic adding wallet id